File: packages/protocol/src/secure/binding.ts
import { concatBytes, equalBytes, int32LE, int64LE, readInt32LE, readInt64LE, uint32LE } from "./bytes.js"
import { aesIgeDecrypt, aesIgeEncrypt, authKeyId, sha1Digest } from "./crypto.js"
import { TlReader, encodeTlBytes } from "./tl.js"

export const BindingConstructor = {
  bindAuthKeyInner: 0x75a3f765,
  bindTempAuthKey: 0xcdd42a05,
} as const

export interface BindAuthKeyInner {
  nonce: bigint
  temporaryAuthKeyId: bigint
  permanentAuthKeyId: bigint
  temporarySessionId: bigint
  expiresAt: number
}

export interface BindTempAuthKeyRequest {
  permanentAuthKeyId: bigint
  nonce: bigint
  expiresAt: number
  encryptedMessage: Uint8Array
}

const keyIdLong = (key: Uint8Array): bigint =>
  new DataView(authKeyId(key).buffer).getBigInt64(0, true)

export const encodeBindAuthKeyInner = (value: BindAuthKeyInner): Uint8Array => concatBytes(
  uint32LE(BindingConstructor.bindAuthKeyInner),
  int64LE(value.nonce),
  int64LE(value.temporaryAuthKeyId),
  int64LE(value.permanentAuthKeyId),
  int64LE(value.temporarySessionId),
  int32LE(value.expiresAt),
)

export const decodeBindAuthKeyInner = (body: Uint8Array): BindAuthKeyInner => {
  if (body.length !== 40 || (readInt32LE(body, 0) >>> 0) !== BindingConstructor.bindAuthKeyInner) {
    throw new RangeError("Invalid bind_auth_key_inner")
  }
  return {
    nonce: readInt64LE(body, 4),
    temporaryAuthKeyId: readInt64LE(body, 12),
    permanentAuthKeyId: readInt64LE(body, 20),
    temporarySessionId: readInt64LE(body, 28),
    expiresAt: readInt32LE(body, 36),
  }
}

export const encodeBindTempAuthKey = (value: BindTempAuthKeyRequest): Uint8Array => {
  if (value.encryptedMessage.length !== 104) throw new RangeError("Binding proof must be exactly 104 bytes")
  return concatBytes(
    uint32LE(BindingConstructor.bindTempAuthKey),
    int64LE(value.permanentAuthKeyId),
    int64LE(value.nonce),
    int32LE(value.expiresAt),
    encodeTlBytes(value.encryptedMessage),
  )
}

export const decodeBindTempAuthKey = (body: Uint8Array): BindTempAuthKeyRequest => {
  if (body.length < 4 || (readInt32LE(body, 0) >>> 0) !== BindingConstructor.bindTempAuthKey) {
    throw new RangeError("Invalid auth.bindTempAuthKey")
  }
  const reader = new TlReader(body.slice(4))
  const value = {
    permanentAuthKeyId: reader.readLong(),
    nonce: reader.readLong(),
    expiresAt: reader.readInt(),
    encryptedMessage: reader.readBytes(),
  }
  reader.expectEnd()
  if (value.encryptedMessage.length !== 104) throw new RangeError("Binding proof must be exactly 104 bytes")
  return value
}

const deriveV1Aes = (authKey: Uint8Array, messageKey: Uint8Array): { key: Uint8Array; iv: Uint8Array } => {
  if (authKey.length !== 256 || messageKey.length !== 16) throw new RangeError("Invalid binding key material")
  const a = sha1Digest(messageKey, authKey.slice(0, 32))
  const b = sha1Digest(authKey.slice(32, 48), messageKey, authKey.slice(48, 64))
  const c = sha1Digest(authKey.slice(64, 96), messageKey)
  const d = sha1Digest(messageKey, authKey.slice(96, 128))
  return {
    key: concatBytes(a.slice(0, 8), b.slice(8, 20), c.slice(4, 16)),
    iv: concatBytes(a.slice(8, 20), b.slice(0, 8), c.slice(16, 20), d.slice(0, 8)),
  }
}

export const createTemporaryKeyBindingProof = (input: {
  permanentAuthKey: Uint8Array
  temporaryAuthKey: Uint8Array
  temporarySessionId: bigint
  messageId: bigint
  nonce: bigint
  expiresAt: number
  randomInt128: Uint8Array
  randomPadding: Uint8Array
}): Uint8Array => {
  if (input.randomInt128.length !== 16 || input.randomPadding.length !== 8) {
    throw new RangeError("Binding proof requires 16-byte prefix and 8-byte padding")
  }
  const inner = encodeBindAuthKeyInner({
    nonce: input.nonce,
    temporaryAuthKeyId: keyIdLong(input.temporaryAuthKey),
    permanentAuthKeyId: keyIdLong(input.permanentAuthKey),
    temporarySessionId: input.temporarySessionId,
    expiresAt: input.expiresAt,
  })
  const plaintextWithoutPadding = concatBytes(
    input.randomInt128, int64LE(input.messageId), int32LE(0), int32LE(inner.length), inner,
  )
  const messageKey = sha1Digest(plaintextWithoutPadding).slice(4, 20)
  const { key, iv } = deriveV1Aes(input.permanentAuthKey, messageKey)
  return concatBytes(
    authKeyId(input.permanentAuthKey),
    messageKey,
    aesIgeEncrypt(concatBytes(plaintextWithoutPadding, input.randomPadding), key, iv),
  )
}

export interface VerifiedTemporaryKeyBinding {
  inner: BindAuthKeyInner
  messageId: bigint
}

export const verifyTemporaryKeyBindingProof = (input: {
  encryptedMessage: Uint8Array
  permanentAuthKey: Uint8Array
  outerPermanentAuthKeyId: bigint
  outerTemporaryAuthKeyId: bigint
  outerTemporarySessionId: bigint
  outerMessageId: bigint
  outerNonce: bigint
  outerExpiresAt: number
  temporaryKeyExpiresAt: number
  nowSeconds: number
}): VerifiedTemporaryKeyBinding => {
  if (input.encryptedMessage.length !== 104 || input.permanentAuthKey.length !== 256) {
    throw new RangeError("Invalid temporary-key binding proof")
  }
  const receivedKeyId = input.encryptedMessage.slice(0, 8)
  const messageKey = input.encryptedMessage.slice(8, 24)
  const { key, iv } = deriveV1Aes(input.permanentAuthKey, messageKey)
  const plaintext = aesIgeDecrypt(input.encryptedMessage.slice(24), key, iv)
  const plaintextWithoutPadding = plaintext.slice(0, 72)
  if (!equalBytes(receivedKeyId, authKeyId(input.permanentAuthKey)) ||
      !equalBytes(messageKey, sha1Digest(plaintextWithoutPadding).slice(4, 20))) {
    throw new RangeError("Invalid temporary-key binding confirmation")
  }
  const messageId = readInt64LE(plaintext, 16)
  const sequenceNumber = readInt32LE(plaintext, 24)
  const bodyLength = readInt32LE(plaintext, 28)
  if (sequenceNumber !== 0 || bodyLength !== 40 || plaintext.length !== 80) {
    throw new RangeError("Invalid isolated binding record")
  }
  const inner = decodeBindAuthKeyInner(plaintext.slice(32, 72))
  if (
    keyIdLong(input.permanentAuthKey) !== input.outerPermanentAuthKeyId ||
    inner.permanentAuthKeyId !== input.outerPermanentAuthKeyId ||
    inner.temporaryAuthKeyId !== input.outerTemporaryAuthKeyId ||
    inner.temporarySessionId !== input.outerTemporarySessionId ||
    inner.nonce !== input.outerNonce ||
    inner.expiresAt !== input.outerExpiresAt ||
    messageId !== input.outerMessageId ||
    input.outerExpiresAt <= input.nowSeconds ||
    input.outerExpiresAt > input.temporaryKeyExpiresAt + 30
  ) throw new RangeError("Temporary-key binding mismatch")
  return { inner, messageId }
}

File: crates/protocol/src/secure/binding.rs
//! Telegram-compatible temporary authorization-key binding.

use super::{InvalidEncryptedRecord, aes_ige_encrypt, auth_key_id};
use sha1::{Digest, Sha1};

const BIND_AUTH_KEY_INNER: u32 = 0x75a3_f765;
const BIND_TEMP_AUTH_KEY: u32 = 0xcdd4_2a05;

fn sha1(parts: &[&[u8]]) -> [u8; 20] {
    let mut digest = Sha1::new();
    for part in parts {
        digest.update(part);
    }
    digest.finalize().into()
}

fn concat(parts: &[&[u8]]) -> Vec<u8> {
    let mut output = Vec::with_capacity(parts.iter().map(|part| part.len()).sum());
    for part in parts {
        output.extend_from_slice(part);
    }
    output
}

fn key_id_long(key: &[u8]) -> Result<i64, InvalidEncryptedRecord> {
    Ok(i64::from_le_bytes(auth_key_id(key)?))
}

fn encode_tl_bytes(value: &[u8]) -> Result<Vec<u8>, InvalidEncryptedRecord> {
    if value.len() > 0x00ff_ffff {
        return Err(InvalidEncryptedRecord);
    }
    let mut output = Vec::new();
    if value.len() < 254 {
        output.push(value.len() as u8);
    } else {
        output.extend_from_slice(&[
            254,
            value.len() as u8,
            (value.len() >> 8) as u8,
            (value.len() >> 16) as u8,
        ]);
    }
    output.extend_from_slice(value);
    output.resize(output.len().next_multiple_of(4), 0);
    Ok(output)
}

fn derive_v1_aes(
    auth_key: &[u8],
    message_key: &[u8],
) -> Result<([u8; 32], [u8; 32]), InvalidEncryptedRecord> {
    if auth_key.len() != 256 || message_key.len() != 16 {
        return Err(InvalidEncryptedRecord);
    }
    let a = sha1(&[message_key, &auth_key[..32]]);
    let b = sha1(&[&auth_key[32..48], message_key, &auth_key[48..64]]);
    let c = sha1(&[&auth_key[64..96], message_key]);
    let d = sha1(&[message_key, &auth_key[96..128]]);
    let key = concat(&[&a[..8], &b[8..20], &c[4..16]]);
    let iv = concat(&[&a[8..20], &b[..8], &c[16..20], &d[..8]]);
    Ok((
        key.try_into().map_err(|_| InvalidEncryptedRecord)?,
        iv.try_into().map_err(|_| InvalidEncryptedRecord)?,
    ))
}

/// Creates the exact isolated encrypted binding proof.
#[allow(clippy::too_many_arguments)]
pub fn create_temporary_key_binding_proof(
    permanent_auth_key: &[u8],
    temporary_auth_key: &[u8],
    temporary_session_id: i64,
    message_id: i64,
    nonce: i64,
    expires_at: i32,
    random_int128: &[u8],
    random_padding: &[u8],
) -> Result<Vec<u8>, InvalidEncryptedRecord> {
    if random_int128.len() != 16 || random_padding.len() != 8 {
        return Err(InvalidEncryptedRecord);
    }
    let inner = concat(&[
        &BIND_AUTH_KEY_INNER.to_le_bytes(),
        &nonce.to_le_bytes(),
        &key_id_long(temporary_auth_key)?.to_le_bytes(),
        &key_id_long(permanent_auth_key)?.to_le_bytes(),
        &temporary_session_id.to_le_bytes(),
        &expires_at.to_le_bytes(),
    ]);
    let plaintext_without_padding = concat(&[
        random_int128,
        &message_id.to_le_bytes(),
        &0_i32.to_le_bytes(),
        &(inner.len() as i32).to_le_bytes(),
        &inner,
    ]);
    let message_key = sha1(&[&plaintext_without_padding])[4..20].to_vec();
    let (key, iv) = derive_v1_aes(permanent_auth_key, &message_key)?;
    Ok(concat(&[
        &auth_key_id(permanent_auth_key)?,
        &message_key,
        &aes_ige_encrypt(
            &concat(&[&plaintext_without_padding, random_padding]),
            &key,
            &iv,
        )?,
    ]))
}

/// Encodes `auth.bindTempAuthKey` around a 104-byte binding proof.
pub fn encode_bind_temporary_auth_key(
    permanent_auth_key_id: i64,
    nonce: i64,
    expires_at: i32,
    proof: &[u8],
) -> Result<Vec<u8>, InvalidEncryptedRecord> {
    if proof.len() != 104 {
        return Err(InvalidEncryptedRecord);
    }
    Ok(concat(&[
        &BIND_TEMP_AUTH_KEY.to_le_bytes(),
        &permanent_auth_key_id.to_le_bytes(),
        &nonce.to_le_bytes(),
        &expires_at.to_le_bytes(),
        &encode_tl_bytes(proof)?,
    ]))
}
